src/core/algorithms/gwo.cpp

```diff
@@ -0,0 +1,107 @@
+#include <stdexcept>
+#include <core/algorithms/gwo.h>
+
+namespace dnn_opt
+{
+namespace core
+{
+namespace algorithms
+{
+
+void gwo::optimize()
+{
+  for (int i = 0; i < get_solutions()->size(); i++)
+  {
+    _generator->generate(2*_dim,_r1);
+    _generator->generate(2*_dim,_r2);
+    auto* current = get_solutions()->get(i)->get_params();
+
+    for (int j = 0; j < _dim; j++)
+    {
+      _A1[j] = 2 * _a * _r1 - _a;
+      _C1[j] = 2 * _r2[j];
+      _Da[j] = abs(_C1[j] * _alpha->get_params()[j] - current[i]);
+      _X1[j] = _alpha->get_params()[j] - _A1[j] * _Da[j];
+      _A1[j] = 2 * _a * _r1 - _a;
+      _C1[j] = 2 * _r2[j];
+      _Db[j] = abs(_C1[j] * _beta->get_params()[j] - current[i]);
+      _X2[j] = _beta->get_params()[j] - _A1[j] * _Db[j];
+      _A1[j] = 2 * _a * _r1 - _a;
+      _C1[j] = 2 * _r2[j];
+      _Dd[j] = abs(_C1[j] * _delta->get_params()[j] - current[i]);
+      _X3[j] = _delta->get_params()[j] - _A1[j] * _Da[j];
```

Comment:
According to the algorithm description, `A` and `C` vectors are generated for each iteration, not for each solution. Hence, it should be done outside this loop.